### PR TITLE
Step 1 of 3: Add opt-in fixing Dialog border radius to match Material Spec

### PR DESCRIPTION
**Pull request by Piinks**
## Description

This PR introduces a temporary opt-in flag as the first step of migrating `Dialog`s to have the correct border radius according to material spec, which has been 2.0 instead of 4.0. Making the new change opt-in allows developers the opportunity to experience the changes ahead of time so that they can fix any issues with the app resulting from the change or fix any broken tests that do not expect details in the new implementation.

Note: The `DatePicker` is already overriding the 2.0 border radius with the correct value (4.0), so it will not be affected.

### Example migration steps

1. Set all `Dialog`, `AlertDialog`, `SimpleDialog` instances, or calls to `showTimePicker` in your app to have `useMaterialBorderRadius` be true. 
```dart
Dialog(
  useMaterialBorderRadius: true, // add this line
  // ...
)
```

2. Examine all screens with Dialogs, AlertDialogs, SimpleDialogs and TimePickers to ensure that they appear as expected, and making sure existing unit tests are not breaking. If something is broken, fix it. An example of a test that might fail is if a unit test was using a Dialog and expecting the original default shape. If the issue cannot be fixed, it may not be an intended change, so file an issue to `flutter/flutter` on GitHub. 

3. A new PR will be introduced in the future to set `Dialog.useMaterialBorderRadius` to `true` by default. This will default everybody's Dialogs to the latest design by default. If developers have properly migrated in step 2, this shouldn't cause any issues to your app. This would be the opportunity to remove the `useMaterialBorderRadius` parameter from your application since it is true by default now and it will be removed in a subsequent PR.
```dart
Dialog(
  // useMaterialBorderRadius: true, // remove this line, since it should be true by default at this point.
  // ...
)
```

4. Finally, a final PR will remove the `useMaterialBorderRadius` parameter altogether. If you removed the `useMaterialBorderRadius` parameter from your `Dialog` instances in step 3, this should not affect your application.

## Tests

Updated remaining test affected by the future default behavior.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. _Non-breaking phased plan_
   - [x] I got input from the developer relations team 
   - [x] I wrote a migration guide, for after we get to step 3 of 3

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by Piinks** [comment]
FYI @darrenaustin

**Comment by Piinks** [comment]
This is ready for review. CI is failing because I have not tagged the deprecation notices with a build, but I'll do that before this lands.

File: packages/flutter/lib/src/material/pickers/date_picker_dialog.dart [hunk 1]
```diff
@@ -487,6 +487,7 @@ class _DatePickerDialogState extends State<_DatePickerDialog> {
         ),
       ),
       insetPadding: const EdgeInsets.symmetric(horizontal: 16.0, vertical: 24.0),
+      // TODO(Piinks): remove once border radius migration is complete
```
Comment (darrenaustin) [inline packages/flutter/lib/src/material/pickers/date_picker_dialog.dart:490]: There is also one of these in the new `date_range_picker_dialog.dart` (line 410). I can update it if you don't want to do it with this PR.

**Review by darrenaustin — COMMENTED** [review]
This all looks great. Thanks so much for taking this on.

Rather than adding several `useMaterialBorderRadius` parameters to the various constructors, would it make sense to just add the interim `useMaterialBorderRadius` parameter to `DialogTheme`? That might reduce the API churn a bit as it would only need to be added and eventually removed from one place.

File: packages/flutter/lib/src/material/dialog.dart [hunk 2]
```diff
@@ -151,7 +166,11 @@ class Dialog extends StatelessWidget {
             child: Material(
               color: backgroundColor ?? dialogTheme.backgroundColor ?? Theme.of(context).dialogBackgroundColor,
               elevation: elevation ?? dialogTheme.elevation ?? _defaultElevation,
-              shape: shape ?? dialogTheme.shape ?? _defaultDialogShape,
+              shape: shape ?? dialogTheme.shape ?? (
```
Comment (Piinks) [inline packages/flutter/lib/src/material/dialog.dart:169]: @darrenaustin I thought I might use a flag via DialogTheme, but the _defaultDialogShape here is actually used as a fallback for the dialogTheme here in the base Dialog class.

I anticipate the turn over being quick, as this is not a severely breaking change. Once this lands I can migrate those affected from the dev roll and remove this flag quickly, the flag itself should not reach stable, instead there will be a migration guide for those that may notice a slight difference in the border radius, or have a golden file test that is affected.

File: packages/flutter/lib/src/material/pickers/date_picker_dialog.dart (hunk 1 above)
Comment (Piinks) [reply packages/flutter/lib/src/material/pickers/date_picker_dialog.dart:490]: Oh thanks! I missed that one! I've added it. :)

Comment (darrenaustin) [reply packages/flutter/lib/src/material/pickers/date_picker_dialog.dart:510]: Nope, you didn't miss it, it was landed after this PR was created. Thanks.

File: packages/flutter/lib/src/material/dialog.dart (hunk 2 above)
Comment (darrenaustin) [reply packages/flutter/lib/src/material/dialog.dart:169]: Ah, that totally makes sense and I agree that it isn't that much churn.

**Review by darrenaustin — APPROVED** [review]
LGTM, thanks.

**Pull request merged**